[Elixir] Concept exercise: basics (#1118)
exercism/v3: languages/elixir/exercises/concept/basics/.docs/introduction.md

> Elixir is a dynamically-typed language, meaning that the type of a variable is only checked at run-time. There is also no real concept of variable assignment.  Using the match `=` operator, we can bind a value to a variable name:

Reviewer: ```suggestion
Elixir is a dynamically-typed language, meaning that the type of a variable is only checked at run-time. Variable names can be _bound_ to a value using the match `=` operator:
```

Author: I thought I had fixed that, sorry, will do asap

Author: this review is based off of an outdated version, it had been fixed

```elixir
variable = 10 # Bind the integer value 10
```

Re-binding a variable's value is also done performed with the `=` operator. Once defined, a variable's type can change when re-bound.

```elixir
count = 1 # Bind an integer value of 1
count = 2 # Re-bind to the new value of 2

# Compiler does not error when binding a new type to the variable
count = false
```

Elixir is an [functional-programming language][functional-programming] and requires all functions to be defined in a _module_. The `defmodule` keyword is used to define a module. All modules are available to all other modules at run-time and do not require an _access modifier_ to make them visible to other parts of the program.  A _module_ is analogous to a _class_ in other programming languages.

```elixir
defmodule Calculator do
  # ...
end
```

> _Functions_ can be defined inside of modules. Each function can have zero or more parameters. All parameters are dynamically typed, and the return type is not explicitly declared, it is the type of the value returned. An _access modifier_ can be specified for functions, making only desired functions available for use external to the module. In a function, the value of the last line is _implicitly returned_ to the calling function.

Reviewer 2: The previous paragraph states:

> [...] requires all functions to be defined in a _module_

Here it states that they _can_ be defined inside of modules. Which is correct?

Author: *named functions, will specify in the doc

Reviewer 2: Same change should be made to the `after.md` document :)

Author: done!

```elixir
defmodule Calculator do
  # ...
end
```

_Named Functions_ must be defined in a module. Each function can have zero or more parameters. All parameters are dynamically typed, and the return type is not explicitly declared, it is the type of the value returned. An _access modifier_ can be specified for functions, making only desired functions available for use external to the module. In a function, the value of the last line is _implicitly returned_ to the calling function.

```elixir
defmodule Calculator do
  def add(x, y) do
    x + y
  end
end
```

Invoking a function is done by specifying its module- and function name and passing arguments for each of the functions's parameters. The module name may be omitted if the function is invoked inside of the module.

```elixir
sum = Calculator.add(1, 2)
```

> When referring to functions, it is very common to refer to them with their _arity_. The _arity_ of a function is the number or parameters it takes.

Reviewer 2: Could you also add this information to the `after.md` document?

Author: Ah, yes!

```elixir

> # add/3 because this function has an arity of 3

Reviewer 2: Tiny suggestion: maybe re-iterate that the function has three parameters?

```suggestion
# add/3 because this function has three parameters and thus an arity of 3
```

Feel free to ignore if you feel this is overkill.
